From: Richard Acayan <mailingradian@gmail.com>
Date: Wed, 13 Sep 2023 19:32:36 -0400
Subject: [PATCH] drm/panel: samsung-sofef00-bonito: move dsi commands to
 disable function

DSI commands are no longer allowed in the unprepare function. Move it to
a new disable function so that the panel can successfully turn off.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 .../gpu/drm/panel/panel-samsung-sofef00-bonito.c   | 52 ++++++++++------------
 1 file changed, 24 insertions(+), 28 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
index 99fc0e7..e5ff9ef 100644
--- a/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
+++ b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
@@ -60,28 +60,6 @@ static int sofef00_on(struct sofef00 *ctx)
 	return 0;
 }
 
-static int sofef00_off(struct sofef00 *ctx)
-{
-	struct mipi_dsi_device *dsi = ctx->dsi;
-	struct device *dev = &dsi->dev;
-	int ret;
-
-	ret = mipi_dsi_dcs_set_display_off(dsi);
-	if (ret < 0) {
-		dev_err(dev, "Failed to set display off: %d\n", ret);
-		return ret;
-	}
-
-	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
-	if (ret < 0) {
-		dev_err(dev, "Failed to enter sleep mode: %d\n", ret);
-		return ret;
-	}
-	msleep(120);
-
-	return 0;
-}
-
 static int sofef00_prepare(struct drm_panel *panel)
 {
 	struct sofef00 *ctx = to_sofef00(panel);
@@ -104,19 +82,36 @@ static int sofef00_prepare(struct drm_panel *panel)
 	return 0;
 }
 
+static int sofef00_disable(struct drm_panel *panel)
+{
+	struct sofef00 *ctx = to_sofef00(panel);
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	struct device *dev = &dsi->dev;
+	int ret;
+
+	ret = mipi_dsi_dcs_set_display_off(dsi);
+	if (ret < 0) {
+		dev_err(dev, "Failed to set display off: %d\n", ret);
+		return ret;
+	}
+
+	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
+	if (ret < 0) {
+		dev_err(dev, "Failed to enter sleep mode: %d\n", ret);
+		return ret;
+	}
+	msleep(120);
+
+	return 0;
+}
+
 static int sofef00_unprepare(struct drm_panel *panel)
 {
 	struct sofef00 *ctx = to_sofef00(panel);
-	struct device *dev = &ctx->dsi->dev;
-	int ret;
 
 	if (!ctx->prepared)
 		return 0;
 
-	ret = sofef00_off(ctx);
-	if (ret < 0)
-		dev_err(dev, "Failed to un-initialize panel: %d\n", ret);
-
 	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
 
 	ctx->prepared = false;
@@ -158,6 +153,7 @@ static int sofef00_get_modes(struct drm_panel *panel,
 
 static const struct drm_panel_funcs sofef00_panel_funcs = {
 	.prepare = sofef00_prepare,
+	.disable = sofef00_disable,
 	.unprepare = sofef00_unprepare,
 	.get_modes = sofef00_get_modes,
 };
